lib/EarlyFunctionAnalysis/UsedRegisters.h

```diff
@@ -0,0 +1,152 @@
+    return true;
+  }
+
+  [[nodiscard]] bool
+  hasArgReg(MetaAddress Addr, model::Register::Values V) const {
+    return true;
+  }
+};
+
+struct UsedRegistersMFI {
+  using Label = BasicBlockNode *;
+  using GraphType = llvm::Inverse<BasicBlockNode *>;
+  using LatticeElement = LatticeElement;
+  using ExtremalLatticeElement = ExtremalLatticeElement;
+
+  DetectABI &DA;
+
+  explicit UsedRegistersMFI(DetectABI &DA) : DA{ DA } {}
+
+  [[nodiscard]] LatticeElement
+  combineValues(const LatticeElement &E1, const LatticeElement &E2) const {
+    LatticeElement Out;
+    for (const auto &[Address, E1Abi] : E1) {
+      if (E2.count(Address)) {
+        const auto &E2Abi = E2.at(Address);
+        const auto &E1Args = E1Abi.ArgumentRegisters;
+        const auto &E1Ret = E1Abi.ReturnRegisters;
+        const auto &E2Args = E2Abi.ArgumentRegisters;
+        const auto &E2Ret = E2Abi.ReturnRegisters;
+        Out[Address]
+          .ArgumentRegisters = FunctionABI::SUL::combineValues(E1Args, E2Args);
+        Out[Address].ReturnRegisters = FunctionABI::SUL::combineValues(E1Ret,
+                                                                       E2Ret);
+      } else {
+        Out[Address] = E1Abi;
+      }
+    }
+
+    for (const auto &[Address, E2Abi] : E2) {
+      if (!Out.count(Address)) {
+        Out[Address] = E2Abi;
+      }
+    }
+
+    return Out;
+  }
+
+  [[nodiscard]] bool
+  isLessOrEqual(const LatticeElement &Left, const LatticeElement &Right) const {
+    for (const auto &[Address, LeftAbi] : Left) {
+      if (Right.count(Address)) {
+        const auto &RightAbi = Right.at(Address);
+        if (!FunctionABI::SUL::isLessOrEqual(LeftAbi.ArgumentRegisters,
+                                             RightAbi.ArgumentRegisters)) {
+          return false;
+        }
+
+        if (!FunctionABI::SUL::isLessOrEqual(LeftAbi.ReturnRegisters,
+                                             RightAbi.ReturnRegisters)) {
+          return false;
+        }
+      }
+    }
+
+    return true;
+  }
+
+  [[nodiscard]] LatticeElement
+  applyTransferFunction(Label L, const LatticeElement &E2) const {
+    LatticeElement Result = E2;
+    auto Reg = [this](const llvm::GlobalVariable *V) {
+      return model::Register::fromCSVName(V->getName(),
+                                          DA.getBinary()->Architecture());
+    };
+
+    const auto &Addr = L->Address;
+    if (Addr.isInvalid()) {
+      return Result;
+    }
+
+    auto Results = DA.analyzeABI(Addr);
+    for (auto &[CSV, State] : Results.ArgumentsRegisters) {
+      if (State == abi::RegisterState::Yes) {
+        Result[Addr].ArgumentRegisters.insert(Reg(CSV));
+      }
+    }
+    for (auto &[CSV, State] : Results.FinalReturnValuesRegisters) {
+      if (State == abi::RegisterState::YesOrDead) {
+        Result[Addr].ReturnRegisters.insert(Reg(CSV));
+      }
+    }
+
+    return Result;
+  }
+
+  static ExtremalLatticeElement ExtremalLattice;
```

Comment:
It's an *element* of the lattice.

```suggestion
  static ExtremalLatticeElement ExtremalLatticeElement;
```